iOS/MSMapsModules/MSMapsModules/MSMapGeoJsonParser/MSMapGeoJsonParser.m

```diff
@@ -0,0 +1,159 @@
+//
+//  MSGeoJsonParser.h
+//  MSMapsModules
+//
+//  Created by Elizabeth Bartusiak (t-elbart) on 6/12/20.
+//  Copyright © 2020 Microsoft Corporation.
+//  Licensed under the MIT license.
+//
+
+#import "MSMapGeoJsonParser.h"
+#import <MicrosoftMaps/MicrosoftMaps.h>
+
+@implementation MSMapGeoJsonParser {
+  MSMapElementLayer *layer;
+};
+
++ (MSMapElementLayer *)parse:(NSString *)geojson error:(NSError **)error {
+  if (geojson == (id)[NSNull null] || geojson.length == 0) {
+    *error = [NSError
+        errorWithDomain:
+            @"com.microsoft.modules.MSMapsModules.InvalidArgumentError"
+                   code:-100
+               userInfo:@{
+                 NSLocalizedDescriptionKey : @"Input String cannot be null."
+               }];
+    return nil;
+  }
+
+  MSMapGeoJsonParser *instance = [[MSMapGeoJsonParser alloc] init];
+  return [instance internalParse:geojson error:error];
+}
+
+- (MSMapElementLayer *)internalParse:(NSString *)geojson
+                               error:(NSError **)error {
+  layer = [[MSMapElementLayer alloc] init];
+  NSData *jsonData = [geojson dataUsingEncoding:NSUTF8StringEncoding];
+  NSError *jsonError;
+  NSDictionary *jsonObject =
+      [NSJSONSerialization JSONObjectWithData:jsonData
+                                      options:kNilOptions
+                                        error:&jsonError];
+
+  if (jsonObject == nil) {
+    *error = [NSError
+        errorWithDomain:@"com.microsoft.modules.MSMapsModules.JSONError"
+                   code:-200
+               userInfo:@{NSLocalizedDescriptionKey : @"Malformed JSON."}];
+    return nil;
+  }
+
+  NSString *type =
+      [NSString stringWithFormat:@"%@", [jsonObject objectForKey:@"type"]];
+  if ([type isEqualToString:@"FeatureCollection"]) {
+    // TODO: parseFeatureCollection
+  } else {
+    if ([type isEqualToString:@"Feature"]) {
+      // TODO: feature
+    }
+    [self switchToType:jsonObject error:error];
+  }
+
+  return layer;
+}
+
+- (void)switchToType:(NSDictionary *)object error:(NSError **)error {
+  NSString *type =
+      [NSString stringWithFormat:@"%@", [object objectForKey:@"type"]];
+  if ([type isEqualToString:@"Polygon"]) {
+    // TODO: parsePolygon
+  } else if ([type isEqualToString:@"Point"]) {
+    NSArray *coordinates = [self getCoordinates:object error:error];
+    if (coordinates != nil) {
+      [self parsePoint:coordinates error:error];
+    } else {
+      layer = nil;
```

Comment:
This pattern of assigning layer to nil to signal an error feels fragile since now all the call stacks need to be inspected to make sure they're not using layer when it could be set to nil by some call further into the callstack. A better approach might be to always return the partically populated layer and then return nil if error != nil in parse()